merlion/models/forecast/deep_base.py

```diff
@@ -0,0 +1,261 @@
+        config = self.config
+
+        # creating model before the training
+        self._create_model()
+        if config.use_gpu:
+            self.to_gpu()
+        else:
+            self.to_cpu()
+
+        total_dataset = RollingWindowDataset(
+            train_data,
+            n_past=config.n_past,
+            n_future=config.max_forecast_steps,
+            batch_size=config.batch_size,
+            target_seq_index=None,  # have to set None, we use config.target_seq_index later in the training, if not this is a bug
+            ts_encoding=config.ts_encoding,
+            valid_fraction=config.valid_fraction,
+            flatten=False,
+            shuffle=True,
+            validation=False,
+        )
+
+        train_steps = len(total_dataset)
+        logger.info(f"Training steps each epoch: {train_steps}")
+
+        bar = ProgressBar(total=config.num_epochs)
+
+        if config.early_stop_patience:
+            early_stopping = EarlyStopping(patience=config.early_stop_patience)
+
+        # start training
+        for epoch in range(config.num_epochs):
+            train_loss = []
+
+            self.deep_model.train()
+            epoch_time = time.time()
+
+            total_dataset.seed = epoch
+            for i, batch in enumerate(total_dataset):
+                self.optimizer.zero_grad()
+
+                loss, _, _ = self._get_batch_model_loss_and_outputs(self._convert_batch_to_tensors(batch))
+                train_loss.append(loss.item())
+
+                loss.backward()
+                if config.clip_gradient is not None:
+                    torch.nn.utils.clip_grad_norm(self.model.parameters(), config.clip_gradient)
+
+                self.optimizer.step()
+
+            train_loss = np.average(train_loss)
+
+            # set validation flag
+            total_dataset.validation = True
+            _, val_loss = self._get_np_loss_and_prediction(total_dataset)
+            total_dataset.validation = False
+
+            if bar is not None:
+                bar.print(
+                    epoch + 1, prefix="", suffix=f"Train Loss: {train_loss: .4f}, Validation Loss: {val_loss: .4f}"
+                )
+
+            if config.early_stop_patience:
+                early_stopping(val_loss, self.deep_model)
+                if early_stopping.early_stop:
+                    logger.info(f"Early stopping with {config.early_stop_patience} patience")
+                    break
+
+        if config.early_stop_patience:
+            early_stopping.load_best_model(self.deep_model)
+            logger.info(f"Load the best model with validation loss: {early_stopping.val_loss_min: .4f}")
+
+        logger.info("End of the training loop")
+
+        pred, _ = self._get_np_loss_and_prediction(total_dataset)
+
+        # since the model predicts multiple steps, we concatenate all the first steps together
+        return pd.DataFrame(pred[:, 0], columns=train_data.columns), None
```

Comment:
A couple of comments:
1. You need to get the predictions by using a new dataset with `valid_fraction = 0` and `shuffle = False`. This is because we want to get the model's predictions for every train window, in order.
2. The dataframe you return must have a `DatetimeIndex`.